### DISCUSSION TITLE
Wezterm Multiplexer + ulimit issue on macOS resolved

### DISCUSSION THREAD

**Discussion by marshallformula** [wezterm/wezterm#3353] [Q&A]
Not sure if this is the correct place for this, but I ran into this strange issue and spent a lot of time trying to figure it out and wanted to make sure it was documented somewhere in case others ran into it.  (feel free to remove if not helpful)

Essentially the issue was this.  Whenever I ran wezterm via multiplexer (configured as shown [in the documentation](https://wezfurlong.org/wezterm/multiplexing.html#unix-domains)) - I would get several errors when trying to update my Neovim plugins using the plugin manager [Lazy.nvim](https://github.com/folke/lazy.nvim)

I didn't capture the error, but it was something along the lines of `fetch failed - unable to spawn process`.  It seemed to show up for random plugins - never the same ones.

If I tried the same thing in wezterm non-multiplexing or in another terminal (eg. kitty) there was no error.

After much tweaking of settings and much digging through the internet - I found something that suggested it could be caused by the arbitrarily low `ulimit` setting on macOS (256 by default for file descriptors). 

I followed the instructions in [this gist](https://gist.github.com/skylock/0117ec637d468f91260927b43b816eda) to update the `ulimit` settings in mac (also an arbitrarily difficult process for such a simple task) and now I no longer receive the error.

---

My Setup:
- 2021 Macbook Pro M1 Max
- Ventura 13.2.1
- Wezterm 20230321-091724 (nightly)
- Neovim - NVIM v0.9.0-dev-1269+gb02880593-dirty (nightly)

_(I was receiving the error with the standard - non nightly - releases of nvim and wezterm as well)_

**Comment by wez** [comment]
What did you raise the limit to?

**Reply by marshallformula** [reply]
I set the file descriptors limit to 64000 and processes limit to 2048 as suggested in that gist.  No issues since

**Reply by wez** [reply]
I've added two new options to make this a bit easier; they work by raising the soft limits of the process to a minimum desirable value.  The default minimum values for nofile and nproc are 2048 each.  They will *not* reduce the soft limit down from an already higher value.

I didn't want to set them to anything super huge by default because there is a common pattern in some software (eg: certain versions/configurations of python's subprocess, `yum` on CentOS 7 bit me recently, and there are a number of others) that will iterate from `0..soft_limit` to speculatively close random file descriptors before spawning subprocesses.  For sufficiently large values, this can take a noticeable amount of time, or even minutes.

The defaults are larger than the meager macOS defaults, but not too crazy, and it is easy to specify something larger in your config if you wish.

* https://wezfurlong.org/wezterm/config/lua/config/ulimit_nofile.html
* https://wezfurlong.org/wezterm/config/lua/config/ulimit_nproc.html

**Reply by wez** [reply]
(these changes will show up in nightly builds in about an hour)

**Reply by marshallformula** [reply]
Fantastic!  Thanks!
